File: android/app/src/main/java/com/foliole/android/FolioleCompanionSyncPackProviderDefinitions.java
package com.foliole.android;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

final class FolioleCompanionSyncPackProviderDefinitions {
    private static final String ASSET = "companion-sync-pack-provider-definitions.json";
    private final JSONObject value;

    private FolioleCompanionSyncPackProviderDefinitions(JSONObject value) { this.value = value; }

    static FolioleCompanionSyncPackProviderDefinitions load(Context context) throws Exception {
        return new FolioleCompanionSyncPackProviderDefinitions(new JSONObject(FolioleCompanionAssetReader.read(context, ASSET)));
    }

    JSONArray copyStatements() throws Exception { return value.getJSONArray("copyStatements"); }
    String databaseEntry() throws Exception { return value.getString("databaseEntry"); }
    String format() throws Exception { return value.getString("format"); }
    int formatVersion() throws Exception { return value.getInt("formatVersion"); }
    int payloadCopyIndex() throws Exception { return value.getInt("payloadCopyIndex"); }
    JSONArray payloadPlans() throws Exception { return value.getJSONArray("payloadPlans"); }
    JSONArray packSchema() throws Exception { return value.getJSONArray("packSchema"); }
    JSONObject preparedMemberDataPlane() throws Exception { return value.getJSONObject("preparedMemberDataPlane"); }
    JSONObject protocol() throws Exception { return value.getJSONObject("protocol"); }
    int schemaVersion() throws Exception { return value.getInt("schemaVersion"); }
    int stateCopyIndex() throws Exception { return value.getInt("stateCopyIndex"); }
    JSONArray tableNames() throws Exception { return value.getJSONArray("tableNames"); }
}

File: ios/App/App/FolioleCompanionSyncPackProviderDefinitions.swift
import Foundation

struct FolioleCompanionSyncPackProviderDefinitions {
    let value: [String: Any]

    static func load() throws -> Self {
        #if SWIFT_PACKAGE
        let bundle = Bundle.module
        #else
        let bundle = Bundle.main
        #endif
        guard let url = bundle.url(
            forResource: "companion-sync-pack-provider-definitions", withExtension: "json"
        ) else { throw invalid("sync_pack_provider_definitions_missing") }
        let decoded = try JSONSerialization.jsonObject(with: Data(contentsOf: url))
        guard let value = decoded as? [String: Any] else { throw invalid("sync_pack_provider_definitions_invalid") }
        return Self(value: value)
    }

    var compression: String { string("compression") }
    var copyStatements: [String] { value["copyStatements"] as? [String] ?? [] }
    var databaseEntry: String { string("databaseEntry") }
    var format: String { string("format") }
    var formatVersion: Int { value["formatVersion"] as? Int ?? -1 }
    var packSchema: [String] { value["packSchema"] as? [String] ?? [] }
    var payloadCopyIndex: Int { value["payloadCopyIndex"] as? Int ?? -1 }
    var payloadPlans: [[String: Any]] { value["payloadPlans"] as? [[String: Any]] ?? [] }
    var preparedMemberDataPlane: [String: Any] { value["preparedMemberDataPlane"] as? [String: Any] ?? [:] }
    var schemaVersion: Int { value["schemaVersion"] as? Int ?? -1 }
    var stateCopyIndex: Int { value["stateCopyIndex"] as? Int ?? -1 }
    var tableNames: [String] { value["tableNames"] as? [String] ?? [] }

    func validate() throws {
        guard compression == "zlib", !copyStatements.isEmpty, !databaseEntry.isEmpty,
              formatVersion > 0, !packSchema.isEmpty, payloadCopyIndex >= 0,
              schemaVersion > 0, stateCopyIndex >= 0, !tableNames.isEmpty else {
            throw Self.invalid("sync_pack_provider_definitions_invalid")
        }
    }

    private func string(_ key: String) -> String { value[key] as? String ?? "" }
    private static func invalid(_ message: String) -> NSError {
        NSError(domain: "FolioleCompanionSyncPackProviderDefinitions", code: 1,
                userInfo: [NSLocalizedDescriptionKey: message])
    }
}
